### PR TITLE
Register errors concerning section topology in ErrorGuard.

### PR DESCRIPTION
**Pull request by blattms**
That will make those (e.g. EDITNNC in the GRiD section) fatal. In addition we now also check (region) operations on keywords such as
```
MULTIPLY
  'TRANX' 2.0 /
  'TRANY' 4.0 /
/

### CODE REVIEW

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
:flushed: Sometime my pushiness is  a bit embarrassing. :sweat_smile:

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
jenkins build this please

**Review by atgeirr — APPROVED** [review]
Looks good.

**Comment by bska** [comment]
> ```
> MULTIPLY
>   'TRANX' 2.0 /
>   'TRANY' 4.0 /
> /
> ```

In principle this is a valid operation (in the `EDIT` section).  Do you mean to say that we now terminate the run if the input tries to perform this operation?

**Comment by blattms** [comment]
Nobody was talking about the `EDIT` section. Am I missing something? This was about editing keywords in the wrong section (like e.g. `GRID` for these particular keywords)

**Comment by bska** [comment]
> Nobody was talking about the `EDIT` section.

Nobody said anything about any specific section.

> Am I missing something?

We're probably both missing something.  The PR description seemed, to me, to apply to all sections and then you used `MULTIPLY 'TRANX'` as an example of something that would be flagged as invalid.  I took that to mean "always invalid" and made my comment on those grounds.

> This was about editing keywords in the wrong section

Right, I got that.  That's why `MULTIPLY` being universally rejected stood out.

**Comment by atgeirr** [comment]
The new section check is not on the operation keyword (`MULTIPLY`) but on the keyword it operates on (`TRANX`).

File: src/opm/input/eclipse/Parser/Parser.cpp [hunk 1]
```diff
@@ -1582,13 +1585,40 @@ std::vector<std::string> Parser::getAllDeckNames () const {
...
-                        "The keyword '"+curKeywordName+"' is located in the '"+curSectionName
-                        +"' section where it is invalid";
-                    OpmLog::warning(Log::fileMessage(curKeyword.location(), msg) );
-                    deckValid = false;
+                const bool isOperateKeyword =
+                    Fieldprops::keywords::is_oper_keyword(curKeywordName);
+
+                auto checker = [&errorGuard, &deckValid, &parser, curSectionName,
+                                ensureKeywordSectionAffiliation, errorKey]
+                    (const std::string& curKeywordName, const KeywordLocation& location)
```
Comment (bska) [inline src/opm/input/eclipse/Parser/Parser.cpp:1593]: I didn't notice this before, but this parameter (`curKeywordName`) shadows [an outer variable](https://github.com/blattms/opm-common/blob/e5c44baf38555cba8f3417b46050330d03025cdc/src/opm/input/eclipse/Parser/Parser.cpp#L1581).  This, in turn, [generates an additional warning](https://ci.opm-project.org/job/opm-common-PR-builder/5993/gcc/new/).  Please fix.

Comment (blattms) [reply src/opm/input/eclipse/Parser/Parser.cpp:1593]: good catch. see #3624

**Pull request merged**